Draft Tenzir v4.1 blog post (#3476)
tenzir/tenzir: web/blog/tenzir-v4.1/index.md

---
title: Tenzir v4.1
authors: [dominiklohmann]
date: 2023-08-31

> tags: [release, pipelines, operators]

Reviewer: ```suggestion
tags: [release, pipelines, operators, app]
```

---

> After our successful launch of app.tenzir.com of Tenzir v4.0 at Black Hat, [the

Reviewer: ```suggestion
After our successful launch of [app.tenzir.com](https://app.tenzir.com) and Tenzir v4.0 at Black Hat, [the
```

new v4.1 release][github-release] continues with several enhancements based on
early feedback. We bring to you a (i) new mechanism to pause pipelines, (ii) a
new operator to match Sigma rules, (iii) new operators for in-pipeline
(de)compression, and (iv) a revamp of the `show` operator.

[github-release]: https://github.com/tenzir/vast/releases/tag/v4.1.0

![Tenzir v4.1](tenzir-v4.1.excalidraw.svg)

<!--truncate-->

## Pausing Pipelines

Tenzir now supports pausing pipelines. Previously, users were able to deploy and
start pipelines, and to let them run until they completed (or failed) or
manually stopped. With new paused state, we now have the following five possible
states of a pipeline:

- *Created*: The pipeline was created, but never started
- *Running*: The pipeline is processing data
- *Paused*: A running pipeline has been suspended but still retains its
  in-memory state
- *Stopped*: A pipeline is not running and has no in-memory state
- *Failed*: The pipeline stopped unexpectedly

> Pausing suspends a pipeline in an instant, causing its execution to stop. When a

Reviewer: ```suggestion
Pausing suspends a pipeline in an instant, causing its execution to stop. This means the pipeline relinquishes all in-memory state, freeing up resources. When a
```

paused pipeline is started, it resumes right where it left off as opposed to

> restarting all the way from the beginning.

Reviewer: Reading this again, I immediately asked myself about delivery guarantees, as "where it left off" potentially implies strong guarantees. I'm not sure how to alleviate this, just pointing it out.

Author: I don't know how to alleviate this either, but I think this blog post does not promise to be as technically accurate as possible, so it shouldn't matter much.

We're also making use of this feature in the app with a new pause action button.
A new tab bar at the top makes it easy to see all pipeline states at a glance.
These features will be enabled in the coming days in the app. Here's a sneak
preview of a mock from our designer:

![Pause Feature Mock](pause-mock.png)

You can also pause a pipeline through the API. Use the [/pipeline/update

> endpoint][update-endpoint] with the new `pause` action to pause a pipeline. The

Reviewer: ```suggestion
endpoint][update-endpoint] with the new `pause` action to suspend a pipeline. The
```

Too much "pause" otherwise.

`start` action resumes a pipeline that is currently paused.

[update-endpoint]: https://docs.tenzir.com/api#/paths/~1pipeline~1update/post

## Sigma Operator

The experimental [`sigma` operator][sigma-docs] executes [Sigma

> rules][sigma-github] on its input and outputs its matches. The operator can work

Reviewer: ```suggestion
rules][sigma-github] on its input and outputs matching events. The operator can work
```

both on files and on directories of rules. Rule directories may be updated while
the operator is running, so that adding a new rule to an already deployed
pipeline is as simple as dropping a Sigma rule into the configured directory.

Consider the `sigma` operator as one concrete instance of [security content
execution][unified-detection] that we enable live and retrospectively. For
example, you can perform historical matching via `export | sigma` and streaming

> execution over a Kafka source via from `kafka --topic events | sigma`. Now that

Reviewer: ```suggestion
execution over a Kafka source via `kafka --topic events | sigma`. Now that
```

```

[show-docs]: https://docs.tenzir.com/operators/sources/show
[connectors-docs]: https://docs.tenzir.com/connectors
[formats-docs]: https://docs.tenzir.com/formats
[operators-docs]: https://docs.tenzir.com/operators

## Compress and Decompress Operators

The [`compress`][compress-docs] and [`decompress`][decompress-docs] operators
make it easy to read and write compressed data. The operator uses Apache Arrow's
compression utilities under the hood, and transparently supports all options
that Apache Arrow supports for streaming compression. The currently supported
codecs are `brotli`, `bz2`, `gzip`, `lz4`, and `zstd`.

For example, the following pipeline creates a Gzip-compressed NDJSON file that
contains all events that were previously imported at the node:

```
export
| write json --compact-output
| compress gzip
| save file /tmp/backup.json.gz
```

[compress-docs]: https://docs.tenzir.com/operators/transformations/compress
[decompress-docs]: https://docs.tenzir.com/operators/transformations/decompress

## Small Things

We're constantly polishing the node's pipeline execution engine, and improving
the app's usability. Since the last release, we've improved pipeline execution

> to smake low pipelines return their initial result faster, and have improved the

Reviewer 2: > smake

Author: make slow, not sure how that typo happened 🫠 Will fix, thanks!

> to make slow pipelines return their initial result faster, and have improved the

Reviewer: ```suggestion
to make slow pipelines return their first results faster, and have improved the
```
